File: collect_app/src/main/java/org/odk/collect/android/utilities/ToastUtils.java
package org.odk.collect.android.utilities;

import android.view.Gravity;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import org.odk.collect.android.application.Collect;

public class ToastUtils {

    private ToastUtils() {

    }

    public static void showShortToast(String message) {
        showToast(message, Toast.LENGTH_SHORT);
    }

    public static void showShortToast(int messageResource) {
        showToast(messageResource, Toast.LENGTH_SHORT);
    }

    public static void showLongToast(String message) {
        showToast(message, Toast.LENGTH_LONG);
    }

    public static void showLongToast(int messageResource) {
        showToast(messageResource, Toast.LENGTH_LONG);
    }

    private static void showToast(String message, int duration) {
        Toast.makeText(Collect.getInstance(), message, duration).show();
    }

    private static void showToast(int messageResource, int duration) {
        Toast.makeText(Collect.getInstance(), Collect.getInstance().getString(messageResource), duration).show();
    }

    public static void showShortToastInMiddle(int messageResource) {
        showToastInMiddle(Collect.getInstance().getString(messageResource), Toast.LENGTH_SHORT);
    }

    public static void showShortToastInMiddle(String message) {
        showToastInMiddle(message, Toast.LENGTH_SHORT);
    }

    public static void showLongToastInMiddle(int messageResource) {
        showToastInMiddle(Collect.getInstance().getString(messageResource), Toast.LENGTH_LONG);
    }

    private static void showToastInMiddle(String message, int duration) {
        Toast toast = Toast.makeText(Collect.getInstance(), message, duration);
        ViewGroup group = (ViewGroup) toast.getView();
        TextView messageTextView = (TextView) group.getChildAt(0);
        messageTextView.setTextSize(21);
        messageTextView.setGravity(Gravity.CENTER);

        toast.setGravity(Gravity.CENTER, 0, 0);
        toast.show();
    }
}

File: collect_app/src/main/java/org/odk/collect/android/utilities/ToastUtils.kt
package org.odk.collect.android.utilities

import android.content.Context
import android.view.Gravity
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast

object ToastUtils {

    private lateinit var toast: Toast

    @JvmStatic
    fun showShortToast(context: Context, message: String) {
        showToast(context, message)
    }

    @JvmStatic
    fun showShortToast(context: Context, messageResource: Int) {
        showToast(context, TranslationHandler.getString(context, messageResource))
    }

    @JvmStatic
    fun showLongToast(context: Context, message: String) {
        showToast(context, message, Toast.LENGTH_LONG)
    }

    @JvmStatic
    fun showLongToast(context: Context, messageResource: Int) {
        showToast(
            context,
            TranslationHandler.getString(context, messageResource),
            Toast.LENGTH_LONG
        )
    }

    @JvmStatic
    fun showShortToastInMiddle(context: Context, message: String) {
        showToastInMiddle(context, message)
    }

    private fun showToast(context: Context, message: String, duration: Int = Toast.LENGTH_SHORT) {
        hideOldToast()
        toast = Toast.makeText(context, message, duration)
        toast.show()
    }

    private fun showToastInMiddle(
        context: Context,
        message: String,
        duration: Int = Toast.LENGTH_SHORT
    ) {
        hideOldToast()
        toast = Toast.makeText(context, message, duration)
        try {
            val group = toast.view as ViewGroup?
            val messageTextView = group!!.getChildAt(0) as TextView
            messageTextView.textSize = 21f
            messageTextView.gravity = Gravity.CENTER
        } catch (ignored: Exception) {
            // ignored
        }
        toast.setGravity(Gravity.CENTER, 0, 0)
        toast.show()
    }

    private fun hideOldToast() {
        if (::toast.isInitialized) {
            toast.cancel()
        }
    }
}
